### DISCUSSION TITLE
Is Kratos production ready?

### DISCUSSION THREAD

**Discussion by timorusky** [ory/kratos#918] [Q&A] [answered]
Hey, currently we are considering using ORY Kratos for our production systems (Kubernetes, native IOS apps). 
In my opinion, the base of the project is already very advanced and easy to handle. 
Other features that are still missing, such as non-localizable email templates, are more of a minor problem for us. 
However, since we are a European company, points like GDPR compliance are very important for us. This raises the question: Can we simply program such features ourselves (account deletion, admin session management, account export etc.) by manipulating the database or does this interfere too much with the ecosystem? What is your opinion on this?

I'm already a fan of the project, looking forward to see more.

**Comment by aeneasr** [comment, marked as answer]
Hi there, you can find info about software maturity of individual projects here: https://www.ory.sh/docs/ecosystem/versioning

Generally speaking - ORY Kratos can be used in production, it does not have critical bugs, and it will most likely perform much better than any homegrown solution. However, what we can not guarantee, is that the migration process to newer versions will not include some work on your end (e.g. integrations, UI, updating email templates).

**Reply by timorusky** [reply]
Thank you, I was not aware of the page. That's the point, programming an identity management system from scratch is beyond our capabilities. However, Kratos seems to be the best option when it comes to the complexity of the project (other alternatives seem more complex). Possibly we can grow together with the project. I will dig a little deeper and play with the system.

**Reply by TMInnovations** [reply]
Hi timorusky,
I am glad to have found a comrade-in-arms! This week I played around a lot with kratos and found out about some of the limitations that come with it (at least at the state it is right now, I'm sure there's more to come). I'd be open to further communication so that we may benefit from each others experiences in the future. I'm happy to see ORY being chosen over alternatives! BR Thomas

**Reply by uRTLy** [reply]
@TMInnovations could you please elaborate on some of the limitations? I am on the same spot at the moment. Can't really decide.

**Reply by TMInnovations** [reply]
Hi @uRTLy,

My point of view is: As long as you want to use it serverside, it's exactly what you wanna use. 

However, while playing around with kratos I more and more got the feeling that using it client-side (especially inside single page applications) with Angular (for example) is not what the developers intended to make available for the api users. 

For example, you can not get access to the output of whoami on the client side (inside a browser). The access-control-allowedorigin is set to * by kratos, which makes it impossible to fetch() the whoami endpoint with credentials, which would be needed for getting back propper information. However, on the server side it's no problem of course. 

I'd love to hear a statement from @aeneasr about this behavior. Maybe there's a way to come around this by changing from sessions to jwt.

**Reply by aeneasr** [reply]
Sorry but that is not true! There are actually already 5 or more community SPA apps in the wild. There might be some smaller issues or inconsistencies but SPA is a first class citizen for ORY Kratos. What is lacking is probably a good tutorial.

Regarding API selfservice flows - DO NOT use them in SPA, use only browser selfservice flows.

**Reply by TMInnovations** [reply]
Sorry from my side - I can only tell from my experiences with ory/kratos, where, for me, it was not possible to build up a registration or login flow without reloading the registration/login page. Like explained above, I also haven't been able to get a correct response after calling whoami client-side, due to the problems reported above. 

My intention with https://github.com/TMInnovations/testorykratos is to give a working example for the use of kratos in an SPA.

Any information about how to better work with ory/kratos is appreciated. I'll definetely try to go the jwt way and - if that works better than sessions - the repo will be updated.

**Reply by vinckr** [reply]
Haven't gotten around to trying out your repo, but definitely try it out over the holidays @TMInnovations !
I think its great that you have motivation to contribute here and I want to help and support you getting everything up and running :).

**Reply by aeneasr** [reply]
Reloading is required in order to set up the cookies and prevent cross-site scripting and other potential attack vectors. Unfortunately, we have not yet found a way around that. Does that impede the UX too much?

**Reply by TMInnovations** [reply]
Thanks for your reply.

Wouldn't fetch() -ing with credentials and defined serverside access-control-allowed-origin headers be an option? This would avoid redirecting. The UX is okay, but after reloading all the angular services' states are gone - and this makes SPA developers lifes harder ;-)

Maybe we can find solutions inside the ory poject in the future. I'd definitely be down to contribute.

**Reply by aeneasr** [reply]
>  The UX is okay, but after reloading all the angular services' states are gone - and this makes SPA developers lifes harder ;-)

Right, that makes sense of course.

So we already support JSON payloads in the registration submit forms. Only the first request (selfservice init browser) has to be a redirect. But yeah it's on the road map to have a reference implementation (similar to RN or express) in 2021 and iron out all the smaller issues one might face.
